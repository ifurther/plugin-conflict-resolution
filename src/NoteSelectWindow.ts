import JoplinData from "api/JoplinData";
import JoplinViewsDialogs from "api/JoplinViewsDialogs";

export class NoteSelectWindow {
    private joplinDialogs : JoplinViewsDialogs;
    private joplinData : JoplinData;
    private joplinInstallDir : string;
    private fileSystem;

    constructor(joplinDialogs : JoplinViewsDialogs, joplinData : JoplinData, fileSystem, joplinInstallDir : string) {
        this.joplinDialogs = joplinDialogs;
        this.joplinData = joplinData;
        this.fileSystem = fileSystem;
        this.joplinInstallDir = joplinInstallDir;
    }

    private handle : string = "";

    public async init(id : string) {
        this.handle = await this.joplinDialogs.create(id);
        await this.joplinDialogs.addScript(this.handle, "./ui/NoteSelectWindow/index.js");
        await this.joplinDialogs.addScript(this.handle, "./ui/NoteSelectWindow/index.css");
    }

    public async openDialog() : Promise<string> {

        const notesList = await this.getNotes();
        const escapedJSON = JSON.stringify(notesList).replace(/"/g, '&quot;');

        const htmlContents = await this.fileSystem.readFile(this.joplinInstallDir + "/ui/NoteSelectWindow/index.html");
        await this.joplinDialogs.setHtml(this.handle, `
            <input id="notesList" type="hidden" value="${escapedJSON}" />
            ${htmlContents}
        `);

        const result = await this.joplinDialogs.open(this.handle);
        
        if(result.id !== 'ok') {
            return "";
        }

        if(result === undefined || result.formData === undefined || result.formData.noteSelectForm === undefined || result.formData.noteSelectForm.noteSelect === undefined)
            throw new Error("No note was selected!");
        
        return result.formData.noteSelectForm.noteSelect;
    }

    private async getNotes() {
        const notes = await this.joplinData.get(['notes'], { fields: ['parent_id', 'title', 'id'], order_by: 'updated_time', order_dir: 'DESC' });
 
        // TODO: Handle the hasMore attribute!

        let notesList = [];
        for(const note of notes.items) {
            
            let currentParent = note.parent_id;
            let curFolderTitle = "";
            while(currentParent !== "") {
                if(curFolderTitle !== "")
                    curFolderTitle = " / " + curFolderTitle;
                let curFolder = await this.joplinData.get(['folders', currentParent], { fields: ['title', 'parent_id'] });
                curFolderTitle = curFolder.title + curFolderTitle;
                currentParent = curFolder.parent_id;
            }
            
            notesList.push({
                id: note.id,
                title: note.title,
                folderName: curFolderTitle
            });
        }

        return notesList;
    }
}