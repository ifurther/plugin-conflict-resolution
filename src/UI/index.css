.conflictRes-Editor {
    width: 700px;
    height: 500px;
}