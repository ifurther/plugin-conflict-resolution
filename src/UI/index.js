let myCodeMirror;

//window.onload(() => {
    myCodeMirror = CodeMirror.MergeView(document.getElementById('conflictRes-Editor'), {
        origLeft: 'test',
        mode: 'markdown',
        lineNumbers: true,
        connect: 'align',
        value: 'ahmed',
        allowEditingOriginals: true
    });
//});