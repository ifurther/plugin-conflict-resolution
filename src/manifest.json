{
	"manifest_version": 1,
	"id": "joplin-plugin-conflict-resolution",
	"app_min_version": "2.0.9",
	"version": "1.2.0",
	"name": "Conflict Resolution",
	"description": "A plugin that makes process of resolving conflicts easier.",
	"author": "Ahmed Alwasifey",
	"homepage_url": "https://github.com/joplin/plugin-conflict-resolution",
	"repository_url": "https://github.com/joplin/plugin-conflict-resolution",
	"keywords": []
}