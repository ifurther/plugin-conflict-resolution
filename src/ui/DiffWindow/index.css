#joplin-plugin-content {
    height: 100vh;
}

h1 {
    margin: 0;
}

#titlesBar {
    display: flex;
    margin-bottom: 10px;
    justify-content: space-between;
}

#moveTitleDiv {
    display: flex;
    align-items: flex-end;
    width: 8%;
}

#moveTitleButton {
    font-size: large;
    height: 35px;
    width: 100%;
}

.titleInput {
    width: 45%;
}

#titleLeft, #titleRight {
    font-size: large;
    padding: 5px;
    width: 100%;
    box-sizing: border-box;
}

#dialogTitle {
    font-size: 40px;
    text-align: center;
}

#dialogSubtitle {
    font-size: medium;
}

#conflictRes-Editor {
    height: 100%;
    overflow: hidden;
}

#buttonsList {
    margin-bottom: 5px;
    display: flex;
    font-size: medium;
    justify-content: flex-end;
    gap: 5px;
}

/* CODE FOR TOOLTIPS */
[data-tooltip]:before {
    /* needed - do not touch */
    content: attr(data-tooltip);
    position: absolute;
    opacity: 0;
    
    /* customizable */
    transition: all 0.15s ease;
    padding: 10px;
    color: #333;
    border-radius: 10px;
    box-shadow: 2px 2px 1px silver;    
}

[data-tooltip]:hover:before {
    /* needed - do not touch */
    opacity: 1;
    
    /* customizable */
    background: gray;
    margin-top: -50px;
    margin-left: 0px;    
}

[data-tooltip]:not([data-tooltip-persistent]):before {
    pointer-events: none;
}