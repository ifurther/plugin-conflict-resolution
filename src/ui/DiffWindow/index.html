<div id="titlesBar">
    <div class="titleInput" id="titleDivLeft">
        <h1>Remote Note</h1>
        <input disabled id="titleLeft" value="First Title"/>
    </div>
    <div class="titleInput" id="titleDivRight">
        <h1>Local Note</h1>
        <input id="titleRight" value="Second Title"/>
    </div>
</div>
<form name="note">
    <input type="hidden" name="noteContents" id="noteContents" value="tmp"/>
</form>
<div id="conflictRes-Editor"></div>