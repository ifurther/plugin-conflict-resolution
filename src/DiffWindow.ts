import JoplinData from "api/JoplinData";
import JoplinViewsDialogs from "api/JoplinViewsDialogs";
import { NoteSelectWindow } from "./NoteSelectWindow";

export class DiffWindow {
    
    private joplinDialogs : JoplinViewsDialogs;
    private joplinData : JoplinData;
    private joplinInstallDir : string;
    private fileSystem;
    private noteSelectWindow : NoteSelectWindow;

    constructor(joplinDialogs : JoplinViewsDialogs, joplinData : JoplinData, fileSystem, joplinInstallDir : string) {
        this.joplinDialogs = joplinDialogs;
        this.joplinData = joplinData;
        this.joplinInstallDir = joplinInstallDir;
        this.fileSystem = fileSystem;
    }


    private handle: string = "";
    
    public async Init(id : string) {
        this.handle = await this.joplinDialogs.create(id);
		await this.joplinDialogs.addScript(this.handle, './lib/codemirror/lib/codemirror.js');
		await this.joplinDialogs.addScript(this.handle, './lib/codemirror/lib/codemirror.css');
		await this.joplinDialogs.addScript(this.handle, './lib/diff_match_patch/diff_match_patch.js');
		await this.joplinDialogs.addScript(this.handle, './lib/codemirror/addon/merge/merge.css');
		await this.joplinDialogs.addScript(this.handle, './ui/DiffWindow/index.js');
		await this.joplinDialogs.addScript(this.handle, './ui/DiffWindow/index.css');
		await this.joplinDialogs.setButtons(this.handle, [
			{
				id: 'submit',
				title: 'Save'
			},
			{
				id: 'cancel',
				title: 'Cancel'
			}
		]);

        this.noteSelectWindow = new NoteSelectWindow(this.joplinDialogs, this.joplinData, this.fileSystem, this.joplinInstallDir);
        await this.noteSelectWindow.init("dialog-note-select");
    }

    public async OpenWindow(noteId: string, compareWithId : string = "") {
        const localNote = await this.joplinData.get(['notes', noteId], {
            fields: ['is_conflict', 'conflict_original_id', 'body', 'title']
        });
    
        if(localNote.is_conflict === 0) {
            throw new Error('This is not a conflict note.');
        }
    
        if(localNote.conflict_original_id === "" && compareWithId === "") {
            compareWithId = await this.noteSelectWindow.openDialog();
        }
    
        const remoteId = (compareWithId == "" ? localNote.conflict_original_id : compareWithId);

        const remoteNote = await this.joplinData.get(['notes', remoteId], {
            fields: ['body', 'title']
        });
    
        const remoteNoteContent = remoteNote.body.replace(/"/g, '&quot;');
        const localNoteContent = localNote.body.replace(/"/g, '&quot;');
        const remoteNoteTitle = remoteNote.title.replace(/"/g, '&quot;');
        const localNoteTitle = localNote.title.replace(/"/g, '&quot;');
    
        const htmlContents = await this.fileSystem.readFile(this.joplinInstallDir + '/ui/DiffWindow/index.html');

        // These inputs are a simple hack in order to pass data into the WebView.
        await this.joplinDialogs.setHtml(this.handle, `
            <input id="pluginInstallDir" type="hidden" value="${this.joplinInstallDir}"/> 
            <input id="remoteNote" type="hidden" value="${remoteNoteContent}"/> 
            <input id="curNote" type="hidden" value="${localNoteContent}"/> 
            <input id="remoteTitle" type="hidden" value="${remoteNoteTitle}"/> 
            <input id="curTitle" type="hidden" value="${localNoteTitle}"/> 
            ${htmlContents}
        `);
    
        const response = await this.joplinDialogs.open(this.handle);
        return response;
    }
}