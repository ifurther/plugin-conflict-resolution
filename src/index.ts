import joplin from 'api';
import {MenuItemLocation} from 'api/types';
import {DiffWindow} from './DiffWindow';
import {NoteSelectWindow} from './NoteSelectWindow';

let diffWindow : DiffWindow;
let noteSelectWindow : NoteSelectWindow;

/**
 * Opens a Diff Window to resolve a conflict.
 * @param {string[]} noteIds Should contain the note to resolve the conflict for.
 */
async function openDiffWindow(noteIds : string[]) {
    if (noteIds.length !== 1) {
    // Comparing multiple notes not yet supported.
        return;
    }

    const noteId = noteIds[0];
    const localNote = await joplin.data.get(['notes', noteId], {
        fields: ['is_conflict', 'conflict_original_id'],
    });

    if (localNote.is_conflict === 0) {
        return await joplin.views.dialogs.showMessageBox('This is not a conflict note.');
    }

    let compareWithId = localNote.conflict_original_id;
    if (compareWithId === '') {
        compareWithId = await noteSelectWindow.openDialog();
    }

    if (compareWithId === '') {
    // No note was found to compare to and user didn't select a note. Cancel button must have been pressed.
        return;
    }

    try {
        const resp = await diffWindow.openWindow(noteId, compareWithId);
        if (resp === null) {
            // If it's null, the Cancel button was pressed.
            return;
        }

        // The note in the normal folder is actually the note that came from the remote server, AKA the one in compareWithId.
        // So, we must replace the contents of that note. The note at `noteId` exists in conflicts folder and will be removed.
        await joplin.data.put(['notes', compareWithId], null, {
            body: resp.NoteContents,
            title: resp.NoteTitle,
        });

        await joplin.data.delete(['notes', noteId]);

        await joplin.views.dialogs.showMessageBox('Successfully resolved conflict!');
    } catch (ex) {
        joplin.views.dialogs.showMessageBox(ex.message);
    }
}

joplin.plugins.register({
    onStart: async function() {
        diffWindow = new DiffWindow(joplin.views.dialogs, joplin.data, joplin.require('fs-extra'), await joplin.plugins.installationDir());
        await diffWindow.init('conflictResolution-diff-dialog');

        noteSelectWindow = new NoteSelectWindow(joplin.views.dialogs, joplin.data, joplin.require('fs-extra'), await joplin.plugins.installationDir());
        await noteSelectWindow.init('conflictResolution-noteSelect-dialog');

        await joplin.commands.register({
            name: 'resolveConflictsCommand',
            label: 'Resolve Conflict',
            enabledCondition: 'oneNoteSelected && inConflictFolder',
            execute: openDiffWindow,
        });

        joplin.views.menuItems.create('resolveConflictsButton', 'resolveConflictsCommand', MenuItemLocation.NoteListContextMenu);

        console.info('Conflict Resolution Plugin loaded!');
    },
});
