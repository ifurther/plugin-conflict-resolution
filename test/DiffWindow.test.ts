const JoplinViewsDialogs = require('./../api/JoplinViewsDialogs.d.ts');
const JoplinData = require('./../api/JoplinData.d.ts');
import {DiffWindow} from '../src/DiffWindow';

/* JOPLIN API MOCK */
const mockedDialogApi = {
    create: jest.fn(),
    addScript: jest.fn(),
    setHtml: jest.fn(),
    open: jest.fn(),
    setButtons: jest.fn(),
};
jest.mock('./../api/JoplinViewsDialogs.d.ts', () => {
    return jest.fn().mockImplementation(() => {
        return {
            create: mockedDialogApi.create,
            addScript: mockedDialogApi.addScript,
            setHtml: mockedDialogApi.setHtml,
            open: mockedDialogApi.open,
            setButtons: mockedDialogApi.setButtons,
        };
    });
});

const mockedDataApi = {
    get: jest.fn(),
};
jest.mock('./../api/JoplinData.d.ts', () => {
    return jest.fn().mockImplementation(() => {
        return {
            get: mockedDataApi.get,
        };
    });
});

const fs = require('fs');
jest.mock('fs');

describe('DiffWindow', () => {
    const mockedJoplinDialogs = new JoplinViewsDialogs();
    const mockedJoplinData = new JoplinData();

    mockedDialogApi.create.mockResolvedValue('test-dialog-id');
    fs.readFile.mockResolvedValue('<form></form>');

    mockedDialogApi.open.mockResolvedValue({
        id: 'submit',
        formData: {
            note: {
                newNoteContents: 'Final Contents',
                newNoteTitle: 'Final Title',
            },
        },
    });
    mockedDataApi.get.mockImplementation((arr, obj) => {
        if (arr[0] === 'notes') {
            if (arr[1] === '8c0d6c64f37f409ba4edf517bab861b5') {
                return {
                    title: 'Test Title',
                    body: 'Test Content',
                };
            } else if (arr[1] === '673f0783febc4ead8187727781ab3b32') {
                return {
                    title: 'Test Title 2',
                    body: 'Test Content 2',
                };
            }
        }
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should setup the dialog properly', async () => {
        const dlg = new DiffWindow(mockedJoplinDialogs, mockedJoplinData, fs, 'D:');

        await dlg.init('abc123');
        expect(mockedDialogApi.create).toBeCalledWith('abc123');
        expect(mockedDialogApi.addScript).toBeCalledTimes(6);
        expect(mockedDialogApi.addScript.mock.calls).toMatchSnapshot('addScriptCalls');
        expect(mockedDialogApi.setButtons).toBeCalled();
        expect(mockedDialogApi.setButtons).toBeCalledWith('test-dialog-id', [
            {
                id: 'submit',
                title: 'Save',
            },
            {
                id: 'cancel',
                title: 'Cancel',
            },
        ]);
    });

    it('should test the dialog having correct html', async () => {
        const dlg = new DiffWindow(mockedJoplinDialogs, mockedJoplinData, fs, 'D:');
        dlg.init('abc123');

        await dlg.openWindow('8c0d6c64f37f409ba4edf517bab861b5', '673f0783febc4ead8187727781ab3b32');

        // This also checks that all the fields are set correctly, including titles, note contents and the html from index.html.
        expect(mockedDialogApi.setHtml.mock.calls[0]).toMatchSnapshot('dialogHtml');
    });

    it('should test ok result', async () => {
        const dlg = new DiffWindow(mockedJoplinDialogs, mockedJoplinData, fs, 'D:');
        dlg.init('abc123');
        const res = await dlg.openWindow('8c0d6c64f37f409ba4edf517bab861b5', '673f0783febc4ead8187727781ab3b32');
        expect(res).toStrictEqual({
            NoteContents: 'Final Contents',
            NoteTitle: 'Final Title',
        });
    });

    it('should test clicking cancel', async () => {
        const dlg = new DiffWindow(mockedJoplinDialogs, mockedJoplinData, fs, 'D:');
        dlg.init('abc123');
        mockedDialogApi.open.mockResolvedValueOnce({
            id: 'cancel',
        });
        const res = await dlg.openWindow('8c0d6c64f37f409ba4edf517bab861b5', '673f0783febc4ead8187727781ab3b32');
        expect(res).toBe(null);
    });
});
