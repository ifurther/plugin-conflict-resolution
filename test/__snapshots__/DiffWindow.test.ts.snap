// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffWindow should setup the dialog properly: addScriptCalls 1`] = `
Array [
  Array [
    "test-dialog-id",
    "./lib/codemirror/lib/codemirror.js",
  ],
  Array [
    "test-dialog-id",
    "./lib/codemirror/lib/codemirror.css",
  ],
  Array [
    "test-dialog-id",
    "./lib/diff_match_patch/diff_match_patch.js",
  ],
  Array [
    "test-dialog-id",
    "./lib/codemirror/addon/merge/merge.css",
  ],
  Array [
    "test-dialog-id",
    "./ui/DiffWindow/index.js",
  ],
  Array [
    "test-dialog-id",
    "./ui/DiffWindow/index.css",
  ],
]
`;

exports[`DiffWindow should test the dialog having correct html: dialogHtml 1`] = `
Array [
  "test-dialog-id",
  "
            <input id=\\"pluginInstallDir\\" type=\\"hidden\\" value=\\"D:\\"/> 
            <input id=\\"remoteNote\\" type=\\"hidden\\" value=\\"Test Content 2\\"/> 
            <input id=\\"curNote\\" type=\\"hidden\\" value=\\"Test Content\\"/> 
            <input id=\\"remoteTitle\\" type=\\"hidden\\" value=\\"Test Title 2\\"/> 
            <input id=\\"curTitle\\" type=\\"hidden\\" value=\\"Test Title\\"/> 
            <form></form>
        ",
]
`;
